device.go

```diff
@@ -42,31 +50,114 @@ var (
+// while deviceAddr is the address at which the device is attached on the bridge.
+func getDevicePCIAddress(pciID string) (string, error) {
+	tokens := strings.Split(pciID, "/")
+
+	if len(tokens) != 2 {
+		return "", fmt.Errorf("PCI Identifier for device should be of format [bridgeAddr/deviceAddr], got %s", pciID)
+	}
+
+	bridgeID := tokens[0]
+	deviceID := tokens[1]
+
+	// Deduce the complete bridge address based on the bridge address identifier passed
+	// and the fact that bridges are attached on the main bus with function 0.
+	pciBridgeAddr := fmt.Sprintf("0000:00:%s.0", bridgeID)
+
+	// Find out the bus exposed by bridge
+	bridgeBusPath := fmt.Sprintf(pciBusPathFormat, sysBusPrefix, pciBridgeAddr)
+
+	files, err := ioutil.ReadDir(bridgeBusPath)
+	if err != nil {
+		return "", fmt.Errorf("Error with getting bridge pci bus : %s", err)
+	}
+
+	busNum := len(files)
+	if busNum != 1 {
+		return "", fmt.Errorf("Expected an entry for bus in %s, got %d entries instead", bridgeBusPath, busNum)
 	}
-	if err := waitForDevice(device.VmPath, devName, checkUevent); err != nil {
+
+	bus := files[0].Name()
+
+	// Device address is based on the bus of the bridge to which it is attached.
+	// We do not pass devices as multifunction, hence the trailing 0 in the address.
+	pciDeviceAddr := fmt.Sprintf("%s:%s.0", bus, deviceID)
+
+	bridgeDevicePCIAddr := fmt.Sprintf("%s/%s", pciBridgeAddr, pciDeviceAddr)
+	agentLog.WithField("completePCIAddr", bridgeDevicePCIAddr).Info("Fetched PCI address for device")
+
+	return bridgeDevicePCIAddr, nil
+}
+
+func getBlockDeviceNodeName(s *sandbox, pciID string) (string, error) {
+	pciAddr, err := getDevicePCIAddress(pciID)
+	if err != nil {
+		return "", err
+	}
+
+	var devName string
+	var notifyChan chan string
+
+	fieldLogger := agentLog.WithField("pciID", pciID)
+
+	// Check if the PCI identifier is in PCI device map.
+	s.Lock()
+	for key, value := range s.pciDeviceMap {
+		if strings.Contains(key, pciAddr) {
+			devName = value
+			fieldLogger.Info("Device found in pci device map")
+			break
+		}
+	}
+
+	// If device is not found in the device map, hotplug event has not
+	// been received yet, create and add channel to the watchers map.
+	// The key of the watchers map is the device we are interested in.
+	// Note this is done inside the lock, not to miss any events from the
+	// global udev listener.
+	if devName == "" {
+		notifyChan := make(chan string, 1)
+		s.deviceWatchers[pciAddr] = notifyChan
+	}
+	s.Unlock()
+
+	if devName == "" {
+		fieldLogger.Info("Waiting on channel for device notification")
+		select {
+		case devName = <-notifyChan:
+		case <-time.After(time.Duration(timeoutHotplug) * time.Second):
+			return "", grpcStatus.Errorf(codes.DeadlineExceeded,
+				"Timeout reached after %ds waiting for device %s",
+				timeoutHotplug, pciAddr)
+		}
+	}
+
+	return filepath.Join(systemDevPath, devName), nil
+}
+
+func virtioBlkDeviceHandler(device pb.Device, spec *pb.Spec, s *sandbox) error {
```

Comment:
Now `device.Id` for both virtio-scsi and virtio-blk are well formatted addresses. Could you please add some comments above both the two handlers documenting the format of the device id?

Reply:
Have added in comments.